### PR TITLE
Added deterministic ordering for pretty much all queries that lock anything

### PR DESCRIPTION
**Pull request by Dantemss**
This should fix multiple deadlocks

Also added some missing array indices

### CODE REVIEW

File: app/domain/services/ecosystem_matrices_updated/service.rb [hunk 1]
```diff
@@ -2,8 +2,10 @@ class Services::EcosystemMatricesUpdated::Service < Services::ApplicationService
   def process(ecosystem_matrices_updated:)
     relevant_update_uuids = ecosystem_matrices_updated.map { |update| update[:calculation_uuid] }
     EcosystemMatrixUpdate.transaction do
-      ecosystem_matrix_updates_by_uuid = EcosystemMatrixUpdate.where(uuid: relevant_update_uuids)
-                                                              .select(:uuid, :algorithm_names)
+      ecosystem_matrix_updates_by_uuid = EcosystemMatrixUpdate.select(:uuid, :algorithm_names)
+                                                              .where(uuid: relevant_update_uuids)
+                                                              .ordered
+                                                              .lock('FOR NO KEY UPDATE')
                                                               .index_by(&:uuid)
```
Comment (nathanstitt) [inline app/domain/services/ecosystem_matrices_updated/service.rb:9]: do you mean to use `:pluck` here?  That'll return a hash which seems to be how `ecosystem_matrix_updates_by_uuid` is accessed on line 17 below

File: db/migrate/20181002193441_change_exercise_uuids_columns_to_string_array_and_add_indices.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+class ChangeExerciseUuidsColumnsToStringArrayAndAddIndices < ActiveRecord::Migration[5.0]
+  def change
+    change_column :student_clue_calculations, :exercise_uuids, :string, array: true
+    add_index :student_clue_calculations, :exercise_uuids, using: :gin
+
+    change_column :teacher_clue_calculations, :exercise_uuids, :string, array: true
+    add_index :teacher_clue_calculations, :exercise_uuids, using: :gin
+  end
```
Comment (nathanstitt) [inline db/migrate/20181002193441_change_exercise_uuids_columns_to_string_array_and_add_indices.rb:8]: will these take a long time?  Wondering if we need to schedule it with DevOps?

Comment (Dantemss) [reply db/migrate/20181002193441_change_exercise_uuids_columns_to_string_array_and_add_indices.rb:8]: Should be fast, I think. These tables don't have too many records. Something like students times ecosystems they are using.

File: app/domain/services/ecosystem_matrices_updated/service.rb (hunk 1 above)
Comment (Dantemss) [reply app/domain/services/ecosystem_matrices_updated/service.rb:9]: I use select to take only the uuid and algorithm_names. I do want the AR object to create the AlgorithmEcosystemMatrixUpdate (well I could use just the uuid for that but meh).

File: db/migrate/20181002193441_change_exercise_uuids_columns_to_string_array_and_add_indices.rb (hunk 2 above)
Comment (Dantemss) [reply db/migrate/20181002193441_change_exercise_uuids_columns_to_string_array_and_add_indices.rb:8]: Hm ok there might be some downtime. 1 million StudentClueCalculations and 200,000 TeacherClueCalculations in production.

File: app/domain/services/ecosystem_matrices_updated/service.rb (hunk 1 above)
Comment (nathanstitt) [reply app/domain/services/ecosystem_matrices_updated/service.rb:9]: Yeah that's what I thought, but `select` will create an AR object, while `pluck` will return them as a hash.  Or so I've always thought at least..  yep, in IRB on Tutor:

```ruby
[6] pry(main)> Tasks::Models::TaskStep.select([:id, :tasked_id]).last
  Tasks::Models::TaskStep Load (0.4ms)  SELECT  "tasks_task_steps"."id", "tasks_task_steps"."tasked_id" FROM "tasks_task_steps"  ORDER BY "tasks_task_steps"."id" DESC LIMIT 1
=> #<Tasks::Models::TaskStep:0x00007f9fbaea9410 id: 24458, tasked_id: 13000>
[7] pry(main)> Tasks::Models::TaskStep.pluck(:id, :tasked_id).last
   (6.4ms)  SELECT "tasks_task_steps"."id", "tasks_task_steps"."tasked_id" FROM "tasks_task_steps"
=> [23829, 7544]
```

Comment (nathanstitt) [reply app/domain/services/ecosystem_matrices_updated/service.rb:9]: and actually I was somewhat inaccurate, I guess pluck returns as nested array, so probably less convenient:

```ruby
[9] pry(main)> Tasks::Models::TaskStep.limit(4).pluck(:id, :tasked_id)
   (0.4ms)  SELECT  "tasks_task_steps"."id", "tasks_task_steps"."tasked_id" FROM "tasks_task_steps" LIMIT 4
=> [[13370, 3641], [13371, 3642], [23631, 7120], [23632, 7121]]
```

**Pull request merged**